plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/GcsTransactionLogSynchronizer.java

```diff
@@ -0,0 +1,126 @@
+        try {
+            createStorageObjectExclusively(newLogEntryPath.toUri(), entryContents, storage);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public boolean isUnsafe()
+    {
+        return false;
+    }
+
+    /**
+     * This method uses {@link Storage} Cloud Storage API client for creating synchronously a storage object
+     * with the specified path and content.
+     * <p>
+     * This approach avoids dealing with `FSDataOutputStream` exposed by the method
+     * `org.apache.hadoop.fs.FileSystem#create(org.apache.hadoop.fs.Path)` and the intricacies of handling
+     * exceptions which may occur while writing the content to the output stream.
+     */
+    private static StorageObject createStorageObjectExclusively(URI blobPath, byte[] content, Storage storage)
+            throws IOException
+    {
+        CreateFileOptions createFileOptions = CreateFileOptions.DEFAULT_NO_OVERWRITE;
+        CreateObjectOptions createObjectOptions = objectOptionsFromFileOptions(createFileOptions);
+        StorageResourceId storageResourceId = StorageResourceId.fromUriPath(blobPath, false);
+
+        StorageObject object =
+                new StorageObject()
+                        .setContentEncoding(createObjectOptions.getContentEncoding())
+                        .setMetadata(encodeMetadata(createObjectOptions.getMetadata()))
+                        .setName(storageResourceId.getObjectName());
+
+        InputStream inputStream = new ByteArrayInputStream(content, 0, content.length);
+        Storage.Objects.Insert insert = storage.objects().insert(
+                storageResourceId.getBucketName(),
+                object,
+                new InputStreamContent(createObjectOptions.getContentType(), inputStream));
+        // By setting `ifGenerationMatch` setting to `0`, the creation of the blob will succeed only
+        // if there are no live versions of the object. When the blob already exists, the operation
+        // will fail with the exception message `412 Precondition Failed`
+        insert.setIfGenerationMatch(0L);
+        insert.getMediaHttpUploader().setDirectUploadEnabled(true);
+        insert.setName(storageResourceId.getObjectName());
+        return insert.execute();
+    }
+
+    /**
+     * Helper for converting from a {@code Map<String, byte[]>} metadata map that may be in a
+     * StorageObject into a {@code Map<String, String>} suitable for placement inside a
+     * {@link GoogleCloudStorageItemInfo}.
+     */
+    private static Map<String, String> encodeMetadata(Map<String, byte[]> metadata)
+    {
+        return Maps.transformValues(metadata, GcsTransactionLogSynchronizer::encodeMetadataValues);
+    }
+
+    private static String encodeMetadataValues(byte[] bytes)
+    {
+        return bytes == null ? Data.NULL_STRING : Base64.getEncoder().encodeToString(bytes);
```

Comment:
Is this dead code?

CreateFileOptions.DEFAULT_NO_OVERWRITE should have no attributes,
so `CreateObjectOptions createObjectOptions` should have no metadata,
so this line should never be called